### PR TITLE
[Bug fix] Reject VM defined with volume with no matching disk

### PR DESCRIPTION
**Pull request by iholder-redhat**
**What this PR does / why we need it**:
Reject VM defined with volume with no matching disk.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Bugzilla bug: https://bugzilla.redhat.com/show_bug.cgi?id=1954667
Github issue: https://github.com/kubevirt/kubevirt/issues/5556

**Release note**:

```release-note
[bugfix] - reject VM defined with volume with no matching disk
```

### CODE REVIEW

File: pkg/virt-api/webhooks/validating-webhook/admitters/vmi-create-admitter.go [hunk 1]
```diff
@@ -901,7 +903,22 @@ func validateBootOrder(field *k8sfield.Path, spec *v1.VirtualMachineInstanceSpec
 			}
 			bootOrderMap[order] = true
 		}
+
+		diskNames[disk.Name] = nil
 	}
+
+	// Validate that volumes match disks correctly
+	for idx, volume := range spec.Volumes {
```
Comment (stu-gott) [inline pkg/virt-api/webhooks/validating-webhook/admitters/vmi-create-admitter.go:911]: OK! this ensures there's always a disk if a volume exists.

Is the reflexive case required? Do we need a volume if a disk exists?

**Comment by stu-gott** [comment]
This appears to be something we could explicitly test for at the functional test level as well.

File: pkg/virt-api/webhooks/validating-webhook/admitters/vmi-create-admitter.go (hunk 1 above)
Comment (iholder-redhat) [reply pkg/virt-api/webhooks/validating-webhook/admitters/vmi-create-admitter.go:911]: The reflexive case is already being checked a few lines above (line 877) and there also a test covering it

File: pkg/virt-api/webhooks/validating-webhook/admitters/vmi-create-admitter.go [hunk 2]
```diff
@@ -901,7 +903,22 @@ func validateBootOrder(field *k8sfield.Path, spec *v1.VirtualMachineInstanceSpec
...
+		if _, machingDiskExists := diskNames[volume.Name]; !machingDiskExists {
```
Comment (vladikr) [inline pkg/virt-api/webhooks/validating-webhook/admitters/vmi-create-admitter.go:912]: @iholder-redhat volumes can also match `spec.Domain.Devices.Filesystems`

Comment (vladikr) [reply pkg/virt-api/webhooks/validating-webhook/admitters/vmi-create-admitter.go:912]: see [1] for example.
[1] https://github.com/kubevirt/kubevirt/blob/master/pkg/virt-launcher/virtwrap/converter/converter.go#L1373

**Comment by iholder-redhat** [comment]
> This appears to be something we could explicitly test for at the functional test level as well.

@stu-gott I've added functional tests :)
p.s. do we have a policy / rule of thumb on when to add a functional test?

File: pkg/virt-api/webhooks/validating-webhook/admitters/vmi-create-admitter.go (hunk 2 above)
Comment (iholder-redhat) [reply pkg/virt-api/webhooks/validating-webhook/admitters/vmi-create-admitter.go]: Thanks!
`diskNames` is now `diskAndFilesystemNames` and contain both fs and disk names :)

File: pkg/virt-api/webhooks/validating-webhook/admitters/vmi-create-admitter.go [hunk 3]
```diff
@@ -861,12 +861,14 @@ func validateNetworkHasOnlyOneType(field *k8sfield.Path, cniTypesCount int, caus
 func validateBootOrder(field *k8sfield.Path, spec *v1.VirtualMachineInstanceSpec, volumeNameMap map[string]*v1.Volume) (bootOrderMap map[uint]bool, causes []metav1.StatusCause) {
 	// used to validate uniqueness of boot orders among disks and interfaces
 	bootOrderMap = make(map[uint]bool)
+	// to perform as set of volume / fs names
+	diskAndFilesystemNames := make(map[string]interface{})
```
Comment (zcahana) [inline pkg/virt-api/webhooks/validating-webhook/admitters/vmi-create-admitter.go:865]: nit: I think it's somewhat more idiomatic/common to use a `map[string]struct{}`, and then "signal the presence" of a disk/fs by having a `diskAndFilesystemNames[disk.Name] = struct{}{}`.

File: pkg/virt-api/webhooks/validating-webhook/admitters/vmi-create-admitter.go [hunk 4]
```diff
@@ -901,7 +903,26 @@ func validateBootOrder(field *k8sfield.Path, spec *v1.VirtualMachineInstanceSpec
...
+		diskAndFilesystemNames[disk.Name] = nil
+	}
+
+	for _, fs := range spec.Domain.Devices.Filesystems {
+		diskAndFilesystemNames[fs.Name] = nil
 	}
+
+	// Validate that volumes match disks and filesystems correctly
+	for idx, volume := range spec.Volumes {
+		if _, machingDiskExists := diskAndFilesystemNames[volume.Name]; !machingDiskExists {
```
Comment (zcahana) [inline pkg/virt-api/webhooks/validating-webhook/admitters/vmi-create-admitter.go:916]: typo-nit: `machingDiskExists` --> `matchingDiskExists`

File: pkg/virt-api/webhooks/validating-webhook/admitters/vmi-create-admitter.go (hunk 3 above)
Comment (iholder-redhat) [reply pkg/virt-api/webhooks/validating-webhook/admitters/vmi-create-admitter.go:865]: Cool. I'm still new to Go, so good to know :)

Comment (zcahana) [reply pkg/virt-api/webhooks/validating-webhook/admitters/vmi-create-admitter.go:865]: Here's a related post by the prophet Dave Cheney:
https://dave.cheney.net/2014/03/25/the-empty-struct

**Comment by zcahana** [comment]
It's worthy to mention that for plain Kubernetes pods, it's perfectly legal to define a pod volume but no volumeMount. The justification for that is unclear to me, and still it has been like that for years and there seem to be no open/closed bug about it.

Nevertheless, we can do better than pods. Thanks @iholder-redhat!
/lgtm

**Comment by vladikr** [comment]
Thank @iholder-redhat 
/lgtm
/approve

**Pull request merged**